api/public/public_api/endpoints/wake_word_file.py

```diff
@@ -0,0 +1,169 @@
+            wake_word = self.request_data["wake_word"].replace(" ", "-")
+            wake_word_dir = data_dir.joinpath("wake-word").joinpath(wake_word)
+            wake_word_dir.mkdir(parents=True, exist_ok=True)
+            file_location_repository = TaggingFileLocationRepository(self.db)
+            self._file_location = file_location_repository.ensure_location_exists(
+                server=LOCAL_IP, directory=str(wake_word_dir)
+            )
+
+        return self._file_location
+
+    def post(self, device_id):
+        """
+        Process a HTTP POST request submitting a wake word sample from a device.
+
+        :param device_id: UUID of the device that originated the request.
+        :return:  HTTP response indicating status of the request.
+        """
+        self._authenticate(device_id)
+        self._validate_post_request()
+        account = self._get_account(device_id)
+        file_name = self._save_audio_file(account)
+        self._add_wake_word_file(account, file_name)
+
+        return jsonify("Wake word sample uploaded successfully"), HTTPStatus.OK
+
+    def _validate_post_request(self):
+        """Load the post request into the validation class and perform validations."""
+        upload_request = UploadRequest(
+            dict(
+                wake_word=self.request.form.get("wake_word"),
+                engine=self.request.form.get("engine"),
+                timestamp=self.request.form.get("timestamp"),
+                model=self.request.form.get("model"),
+            )
+        )
+        upload_request.validate()
+        self.request_data = upload_request.to_native()
+        if "audio_file" not in self.request.files:
+            raise DataError(dict(audio_file="No audio file included in request"))
+
+    def _get_account(self, device_id: str):
+        """Use the device ID to find the account.
+
+        :param device_id: The database ID for the device that made this API call
+        """
+        account_repository = AccountRepository(self.db)
+        return account_repository.get_account_by_device_id(device_id)
+
+    def _save_audio_file(self, account: Account):
+        """Build the file path for the audio file.
+
+        :param account: the account from which sample originated
+        """
+        file_name = f"{account.id}.{self.request_data['timestamp']}.wav"
+        file_path = Path(self.file_location.directory).joinpath(file_name)
+        self.request.files["audio_file"].save(str(file_path))
+
+        return file_name
+
+    def _add_wake_word_file(self, account: Account, file_name: str):
+        """Add the sample to the database for reference and classification.
+
+        :param account: the account from which sample originated
+        :param file_name: name of the audio file saved to file system
+        """
+        sample = WakeWordFile(
+            account_id=account.id,
+            location=self.file_location,
+            name=file_name,
+            origin="mycroft",
```

Comment:
What is origin used for? How do we handle multiple devices for the same account?

Reply:
The "origin" column is to know if the sample came from a device, the soon-to-be-built wake word submission tool, or added manually (like background noise samples).

Does it matter which device a sample came from?  Right now we are just tracking at the account level.

Reply:
I don't know. That is a good question. I was just confused because I did not see a value passed in for origin so I was not sure how it would ever be anything but 'mycroft'.